corebehrt/functional/tokenize.py

```diff
@@ -0,0 +1,100 @@
+import dask.dataframe as dd
+
+
+def _get_segment_change(df: dd.DataFrame) -> dd.Series:
+    """Select entries where segment changes."""
+    return (df.segment != df.segment.shift(-1)) | (df.PID != df.PID.shift(-1))
+
+
+def _get_first_event(df: dd.DataFrame) -> dd.Series:
+    """Select the first event in the group based on abspos."""
+    return (df.PID != df.PID.shift(1)).fillna(True)
+
+
+def _add_token(
+    df: dd.DataFrame, token: str, mask_func: callable, abspos_adjustment: float
+) -> dd.DataFrame:
+    """
+    Insert tokens into the DataFrame based on the mask_func applied to the sorted (PID,abspos) dataframe.
+    The abspos_adjustment is used to ensure the tokens will be placed correctly in the final sequences.
+    """
+    df = df.sort_values(["PID", "abspos"])
+    df_change = df[mask_func(df)]
+    df_change["concept"] = token
+    df_change["abspos"] += abspos_adjustment
+    return dd.concat([df, df_change]).sort_values(["PID", "abspos"])
+
+
+def add_separator_token(
+    features: dd.DataFrame, sep_token: str = "[SEP]"
+) -> dd.DataFrame:
+    """Add separator token after each segment in the dataframe"""
+    return _add_token(features, sep_token, _get_segment_change, 1e-3)
+
+
+def add_cls_token(features: dd.DataFrame, cls_token: str = "[CLS]"):
+    """Add a classification token to the beginning of each patient's sequence"""
+    return _add_token(features, cls_token, _get_first_event, -1e-3)
+
+
+def tokenize(
+    features: dd.DataFrame, vocabulary: dict, frozen_vocab: bool, inplace: bool = False
+) -> dd.DataFrame:
+    """
+    Tokenize the concepts in the features DataFrame.
+    if frozen_vocab is True, the vocabulary will not be updated.
+    And the tokens will be replaced with the UNK token.
+    """
+    if frozen_vocab:
+        features = tokenize_frozen(features, vocabulary, inplace=inplace)
+    else:
+        features, vocabulary = tokenize_with_update(
+            features, vocabulary, inplace=inplace
+        )
+
+    return features, vocabulary
+
+
+def tokenize_with_update(
```

Comment:
Same here, should just take concepts rather than features

Reply:
Fixed.

Reply:
Fixed.